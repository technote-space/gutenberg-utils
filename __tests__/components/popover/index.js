/* eslint-disable no-magic-numbers */
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

const { useRef } = wp.element;
import Popover, {
	useThrottledWindowScrollOrResizeFunc,
	refreshAnchorRectFunc,
	intervalRefreshAnchorRectFunc,
	useFocusContentOnMountFunc,
	renderPopover,
} from '../../../src/components/popover';

const { Component } = wp.element;
const { SlotFillProvider, IsolatedEventContainer } = wp.components;

jest.useFakeTimers();

class PopoverWrapper extends Component {
	render() {
		return <Popover { ...this.props } />;
	}
}

describe( 'Popover', () => {
	const onKeyDown = jest.fn();
	const wrapper = mount(
		<SlotFillProvider>
			<PopoverWrapper
				onKeyDown={ onKeyDown }
			>
				Hello
			</PopoverWrapper>
			<Popover.Slot/>
		</SlotFillProvider>,
	);

	beforeEach( () => {
		onKeyDown.mockClear();
		jest.advanceTimersByTime( 1 );
	} );

	afterEach( () => {
		if ( document.activeElement ) {
			document.activeElement.blur();
		}
	} );

	it( 'should render content', () => {
		expect( toJson( wrapper ) ).toMatchSnapshot();
	} );

	it( 'should call onKeyDown', () => {
		wrapper.find( IsolatedEventContainer ).props().onKeyDown( { keyCode: 38 } ); // â†‘
		expect( onKeyDown ).toHaveBeenCalledTimes( 1 );
		expect( onKeyDown ).toHaveBeenCalledWith( { keyCode: 38 } );
	} );
} );

describe( 'Popover', () => {
	const onClose = jest.fn();
	const wrapper = mount(
		<SlotFillProvider>
			<PopoverWrapper
				onClose={ onClose }
			>
				Hello
			</PopoverWrapper>
			<Popover.Slot/>
		</SlotFillProvider>,
	);

	beforeEach( () => {
		onClose.mockClear();
		jest.advanceTimersByTime( 1 );
	} );

	afterEach( () => {
		if ( document.activeElement ) {
			document.activeElement.blur();
		}
	} );

	it( 'should render content', () => {
		expect( toJson( wrapper ) ).toMatchSnapshot();
	} );

	it( 'should call onClose', () => {
		const stopPropagation = jest.fn();
		wrapper.find( IsolatedEventContainer ).props().onKeyDown( {
			keyCode: 27,
			stopPropagation,
		} ); // esc
		expect( onClose ).toHaveBeenCalledTimes( 1 );
		expect( stopPropagation ).toHaveBeenCalledTimes( 1 );
	} );
} );

describe( 'Popover', () => {
	it( 'should render mobile content', () => {
		const width = window.innerWidth;
		window.innerWidth = 600;
		const wrapper = mount(
			<SlotFillProvider>
				<PopoverWrapper
					expandOnMobile={ true }
					position='middle'
				>
					Hello
				</PopoverWrapper>
				<Popover.Slot/>
			</SlotFillProvider>,
		);
		expect( toJson( wrapper ) ).toMatchSnapshot();
		window.innerWidth = width;
	} );
} );

describe( 'useThrottledWindowScrollOrResizeFunc', () => {
	beforeEach( () => {
		jest.spyOn( window, 'requestAnimationFrame' ).mockImplementation( cb => cb() );
	} );

	afterEach( () => {
		window.requestAnimationFrame.mockRestore();
	} );

	it( 'should setup refresh event 1', () => {
		const handler = jest.fn();
		const contains = jest.fn( () => false );
		const remove = useThrottledWindowScrollOrResizeFunc( handler, { current: { contains } } );
		expect( typeof remove ).toBe( 'function' );

		window.dispatchEvent( new Event( 'resize' ) );
		expect( contains ).toHaveBeenCalledTimes( 0 );
		expect( handler ).toHaveBeenCalledTimes( 1 );
		contains.mockClear();
		handler.mockClear();

		window.dispatchEvent( new Event( 'scroll' ) );
		expect( contains ).toHaveBeenCalledTimes( 1 );
		expect( handler ).toHaveBeenCalledTimes( 1 );
		contains.mockClear();
		handler.mockClear();

		remove();
		window.dispatchEvent( new Event( 'scroll' ) );
		expect( contains ).toHaveBeenCalledTimes( 0 );
		expect( handler ).toHaveBeenCalledTimes( 0 );
	} );

	it( 'should setup refresh event 2', () => {
		const handler = jest.fn();
		const contains = jest.fn( () => true );
		const remove = useThrottledWindowScrollOrResizeFunc( handler, { current: { contains } } );
		expect( typeof remove ).toBe( 'function' );

		window.dispatchEvent( new Event( 'scroll' ) );
		expect( contains ).toHaveBeenCalledTimes( 1 );
		expect( handler ).toHaveBeenCalledTimes( 0 );
	} );

	it( 'should setup refresh event 3', () => {
		const handler = jest.fn();
		const contains = jest.fn( () => false );
		const remove = useThrottledWindowScrollOrResizeFunc( handler, null );
		expect( typeof remove ).toBe( 'function' );

		window.dispatchEvent( new Event( 'resize' ) );
		expect( contains ).toHaveBeenCalledTimes( 0 );
		expect( handler ).toHaveBeenCalledTimes( 1 );
		contains.mockClear();
		handler.mockClear();

		window.dispatchEvent( new Event( 'scroll' ) );
		expect( contains ).toHaveBeenCalledTimes( 0 );
		expect( handler ).toHaveBeenCalledTimes( 1 );
	} );
} );

describe( 'refreshAnchorRectFunc', () => {
	it( 'should do nothing 1', () => {
		const getAnchorRect = jest.fn();
		const setAnchor = jest.fn();
		expect( refreshAnchorRectFunc( { current: null }, null, getAnchorRect, {}, setAnchor ) ).toBeFalsy();
		expect( getAnchorRect ).toHaveBeenCalledTimes( 0 );
		expect( setAnchor ).toHaveBeenCalledTimes( 0 );
	} );

	it( 'should do nothing 2', () => {
		const getAnchorRect = jest.fn();
		const setAnchor = jest.fn();
		expect( refreshAnchorRectFunc( { current: {} }, { x: 0 }, getAnchorRect, { x: 0 }, setAnchor ) ).toBeFalsy();
		expect( getAnchorRect ).toHaveBeenCalledTimes( 0 );
		expect( setAnchor ).toHaveBeenCalledTimes( 0 );
	} );

	it( 'should set anchorRect', () => {
		const getAnchorRect = jest.fn();
		const setAnchor = jest.fn();
		expect( refreshAnchorRectFunc( { current: {} }, {
			x: 123,
			y: 456,
		}, getAnchorRect, {}, setAnchor ) ).toBeTruthy();
		expect( getAnchorRect ).toHaveBeenCalledTimes( 0 );
		expect( setAnchor ).toHaveBeenCalledTimes( 1 );
		expect( setAnchor ).toHaveBeenCalledWith( {
			x: 123,
			y: 456,
		} );
	} );

	it( 'should set getAnchorRect', () => {
		const getAnchorRect = jest.fn( () => ( {
			x: 456,
			y: 789,
		} ) );
		const setAnchor = jest.fn();
		expect( refreshAnchorRectFunc( {
			current: {
				test: 'test',
			},
		}, null, getAnchorRect, {}, setAnchor ) ).toBeTruthy();
		expect( getAnchorRect ).toHaveBeenCalledTimes( 1 );
		expect( getAnchorRect ).toHaveBeenCalledWith( { test: 'test' } );
		expect( setAnchor ).toHaveBeenCalledTimes( 1 );
		expect( setAnchor ).toHaveBeenCalledWith( {
			x: 456,
			y: 789,
		} );
	} );
} );

describe( 'intervalRefreshAnchorRectFunc', () => {
	it( 'should setup setInterval', () => {
		const callback = jest.fn();
		const func = intervalRefreshAnchorRectFunc( null, callback );
		expect( typeof func ).toBe( 'function' );
		expect( callback ).toHaveBeenCalledTimes( 0 );
		jest.advanceTimersByTime( 600 );
		expect( callback ).toHaveBeenCalledTimes( 1 );

		func();

		jest.advanceTimersByTime( 600 );
		expect( callback ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'should not setInterval', () => {
		const callback = jest.fn();
		const func = intervalRefreshAnchorRectFunc( {}, callback );
		expect( typeof func ).toBe( 'undefined' );
		expect( callback ).toHaveBeenCalledTimes( 0 );
		jest.advanceTimersByTime( 600 );
		expect( callback ).toHaveBeenCalledTimes( 0 );
	} );
} );

describe( 'useFocusContentOnMountFunc', () => {
	it( 'should do nothing', () => {
		expect( useFocusContentOnMountFunc( null, { current: {} }, {} ) ).toBeFalsy();
		expect( useFocusContentOnMountFunc( 'firstElement', {}, {} ) ).toBeFalsy();
		expect( useFocusContentOnMountFunc( 'test', { current: {} }, {} ) ).toBeFalsy();
	} );

	it( 'should call focus 1', () => {
		const focus = jest.fn();
		expect( useFocusContentOnMountFunc( 'firstElement', {
			current: {},
		}, {
			tabbable: {
				find: () => [
					{ focus },
				],
			},
		} ) ).toBeTruthy();
		expect( focus ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'should call focus 2', () => {
		const focus = jest.fn();
		expect( useFocusContentOnMountFunc( 'firstElement', {
			current: { focus },
		}, {
			tabbable: {
				find: () => [],
			},
		} ) ).toBeTruthy();
		expect( focus ).toHaveBeenCalledTimes( 1 );
	} );

	it( 'should call focus 3', () => {
		const focus = jest.fn();
		expect( useFocusContentOnMountFunc( 'container', {
			current: { focus },
		}, {
			tabbable: {
				find: () => [],
			},
		} ) ).toBeTruthy();
		expect( focus ).toHaveBeenCalledTimes( 1 );
	} );
} );

describe( 'renderPopover', () => {
	it( 'should render popover', () => {
		const anchorRef = useRef( null );
		const contentRef = useRef( null );
		const popoverPosition = {
			popoverLeft: 10,
			popoverTop: 20,
			yAxis: 'middle',
			xAxis: 'center',
			contentHeight: 30,
			contentWidth: 40,
			isMobile: false,
		};

		class WrapperComponent extends Component {
			render() {
				return renderPopover( {
					anchorRef,
					contentRef,
					contentSize: null,
					popoverPosition,
					headerTitle: '',
					expandOnMobile: false,
					contentProps: {},
					children: null,
					onClose: jest.fn(),
					onKeyDown: jest.fn(),
					onClickOutside: jest.fn(),
				} );
			}
		}

		const wrapper = mount(
			<SlotFillProvider>
				<WrapperComponent/>
				<Popover.Slot/>
			</SlotFillProvider>,
		);
		expect( toJson( wrapper ) ).toMatchSnapshot();
	} );

	it( 'should render mobile popover', () => {
		const anchorRef = useRef( null );
		const contentRef = useRef( null );
		const popoverPosition = {
			popoverLeft: 10,
			popoverTop: 20,
			yAxis: 'middle',
			xAxis: 'center',
			contentHeight: 30,
			contentWidth: 40,
			isMobile: true,
		};

		class WrapperComponent extends Component {
			render() {
				return renderPopover( {
					anchorRef,
					contentRef,
					contentSize: null,
					popoverPosition,
					headerTitle: '',
					expandOnMobile: true,
					contentProps: {},
					children: null,
					onClose: jest.fn(),
					onKeyDown: jest.fn(),
					onClickOutside: jest.fn(),
				} );
			}
		}

		const wrapper = mount(
			<SlotFillProvider>
				<WrapperComponent/>
				<Popover.Slot/>
			</SlotFillProvider>,
		);
		expect( toJson( wrapper ) ).toMatchSnapshot();
	} );
} );
