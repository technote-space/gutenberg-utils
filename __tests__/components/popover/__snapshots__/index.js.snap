// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should render content 1`] = `
<SlotFillProvider>
  <PopoverWrapper
    onKeyDown={[MockFunction]}
  >
    <Popover
      onKeyDown={[MockFunction]}
    >
      <span>
        <Fill
          name="Popover"
        >
          <FillComponent
            name="Popover"
            registerFill={[Function]}
            unregisterFill={[Function]}
          >
            <Portal
              containerInfo={
                <div>
                  <div
                    tabindex="-1"
                  >
                    <div>
                      <div
                        class="components-popover is-bottom is-center components-animate__appear is-from-top"
                      >
                        <div
                          class="components-popover__content"
                          tabindex="-1"
                        >
                          Hello
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              }
            >
              <WithConstrainedTabbing(WithFocusReturn(Component))>
                <div
                  onKeyDown={[Function]}
                  tabIndex="-1"
                >
                  <WithFocusReturn(Component)>
                    <FocusReturn
                      childProps={
                        Object {
                          "children": <clickOutside(PopoverDetectOutside)>
                            <Animate
                              options={
                                Object {
                                  "origin": "top center",
                                }
                              }
                              type="appear"
                            >
                              [Function]
                            </Animate>
                          </clickOutside(PopoverDetectOutside)>,
                        }
                      }
                      focus={
                        Object {
                          "focusHistory": Array [],
                        }
                      }
                    >
                      <div
                        onBlur={[Function]}
                        onFocus={[Function]}
                      >
                        <Component>
                          <clickOutside(PopoverDetectOutside)>
                            <PopoverDetectOutside>
                              <Animate
                                options={
                                  Object {
                                    "origin": "top center",
                                  }
                                }
                                type="appear"
                              >
                                <ForwardRef
                                  className="components-popover is-bottom is-center components-animate__appear is-from-top"
                                  onKeyDown={[Function]}
                                  style={
                                    Object {
                                      "left": undefined,
                                      "top": undefined,
                                      "visibility": undefined,
                                    }
                                  }
                                >
                                  <div
                                    className="components-popover is-bottom is-center components-animate__appear is-from-top"
                                    onKeyDown={[Function]}
                                    onMouseDown={[Function]}
                                    style={
                                      Object {
                                        "left": undefined,
                                        "top": undefined,
                                        "visibility": undefined,
                                      }
                                    }
                                  >
                                    <div
                                      className="components-popover__content"
                                      style={
                                        Object {
                                          "maxHeight": undefined,
                                          "maxWidth": undefined,
                                        }
                                      }
                                      tabIndex="-1"
                                    >
                                      Hello
                                    </div>
                                  </div>
                                </ForwardRef>
                              </Animate>
                            </PopoverDetectOutside>
                          </clickOutside(PopoverDetectOutside)>
                        </Component>
                      </div>
                    </FocusReturn>
                  </WithFocusReturn(Component)>
                </div>
              </WithConstrainedTabbing(WithFocusReturn(Component))>
            </Portal>
          </FillComponent>
        </Fill>
      </span>
    </Popover>
  </PopoverWrapper>
  <Component>
    <Slot
      bubblesVirtually={true}
      name="Popover"
    >
      <SlotComponent
        bubblesVirtually={true}
        getFills={[Function]}
        name="Popover"
        registerSlot={[Function]}
        unregisterSlot={[Function]}
      >
        <div />
      </SlotComponent>
    </Slot>
  </Component>
</SlotFillProvider>
`;

exports[`Popover should render content 2`] = `
<SlotFillProvider>
  <PopoverWrapper
    onClose={[MockFunction]}
  >
    <Popover
      onClose={[MockFunction]}
    >
      <span>
        <Fill
          name="Popover"
        >
          <FillComponent
            name="Popover"
            registerFill={[Function]}
            unregisterFill={[Function]}
          >
            <Portal
              containerInfo={
                <div>
                  <div
                    tabindex="-1"
                  >
                    <div>
                      <div
                        class="components-popover is-bottom is-center components-animate__appear is-from-top"
                      >
                        <div
                          class="components-popover__content"
                          tabindex="-1"
                        >
                          Hello
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              }
            >
              <WithConstrainedTabbing(WithFocusReturn(Component))>
                <div
                  onKeyDown={[Function]}
                  tabIndex="-1"
                >
                  <WithFocusReturn(Component)>
                    <FocusReturn
                      childProps={
                        Object {
                          "children": <clickOutside(PopoverDetectOutside)
                            onClickOutside={[MockFunction]}
                          >
                            <Animate
                              options={
                                Object {
                                  "origin": "top center",
                                }
                              }
                              type="appear"
                            >
                              [Function]
                            </Animate>
                          </clickOutside(PopoverDetectOutside)>,
                        }
                      }
                      focus={
                        Object {
                          "focusHistory": Array [],
                        }
                      }
                    >
                      <div
                        onBlur={[Function]}
                        onFocus={[Function]}
                      >
                        <Component>
                          <clickOutside(PopoverDetectOutside)
                            onClickOutside={[MockFunction]}
                          >
                            <PopoverDetectOutside
                              onClickOutside={[MockFunction]}
                            >
                              <Animate
                                options={
                                  Object {
                                    "origin": "top center",
                                  }
                                }
                                type="appear"
                              >
                                <ForwardRef
                                  className="components-popover is-bottom is-center components-animate__appear is-from-top"
                                  onKeyDown={[Function]}
                                  style={
                                    Object {
                                      "left": undefined,
                                      "top": undefined,
                                      "visibility": undefined,
                                    }
                                  }
                                >
                                  <div
                                    className="components-popover is-bottom is-center components-animate__appear is-from-top"
                                    onKeyDown={[Function]}
                                    onMouseDown={[Function]}
                                    style={
                                      Object {
                                        "left": undefined,
                                        "top": undefined,
                                        "visibility": undefined,
                                      }
                                    }
                                  >
                                    <div
                                      className="components-popover__content"
                                      style={
                                        Object {
                                          "maxHeight": undefined,
                                          "maxWidth": undefined,
                                        }
                                      }
                                      tabIndex="-1"
                                    >
                                      Hello
                                    </div>
                                  </div>
                                </ForwardRef>
                              </Animate>
                            </PopoverDetectOutside>
                          </clickOutside(PopoverDetectOutside)>
                        </Component>
                      </div>
                    </FocusReturn>
                  </WithFocusReturn(Component)>
                </div>
              </WithConstrainedTabbing(WithFocusReturn(Component))>
            </Portal>
          </FillComponent>
        </Fill>
      </span>
    </Popover>
  </PopoverWrapper>
  <Component>
    <Slot
      bubblesVirtually={true}
      name="Popover"
    >
      <SlotComponent
        bubblesVirtually={true}
        getFills={[Function]}
        name="Popover"
        registerSlot={[Function]}
        unregisterSlot={[Function]}
      >
        <div />
      </SlotComponent>
    </Slot>
  </Component>
</SlotFillProvider>
`;

exports[`Popover should render mobile content 1`] = `
<SlotFillProvider>
  <PopoverWrapper
    expandOnMobile={true}
    position="middle"
  >
    <Popover
      expandOnMobile={true}
      position="middle"
    >
      <span>
        <Fill
          name="Popover"
        >
          <FillComponent
            name="Popover"
            registerFill={[Function]}
            unregisterFill={[Function]}
          >
            <Portal
              containerInfo={
                <div>
                  <div
                    tabindex="-1"
                  >
                    <div>
                      <div
                        class="components-popover is-bottom is-center is-mobile components-animate__appear is-from-top"
                      >
                        <div
                          class="components-popover__header"
                        >
                          <span
                            class="components-popover__header-title"
                          />
                          <button
                            class="components-button components-popover__close has-icon"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="dashicon dashicons-no-alt"
                              focusable="false"
                              height="20"
                              role="img"
                              viewBox="0 0 20 20"
                              width="20"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <path
                                d="M14.95 6.46L11.41 10l3.54 3.54-1.41 1.41L10 11.42l-3.53 3.53-1.42-1.42L8.58 10 5.05 6.47l1.42-1.42L10 8.58l3.54-3.53z"
                              />
                            </svg>
                          </button>
                        </div>
                        <div
                          class="components-popover__content"
                          tabindex="-1"
                        >
                          Hello
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              }
            >
              <WithConstrainedTabbing(WithFocusReturn(Component))>
                <div
                  onKeyDown={[Function]}
                  tabIndex="-1"
                >
                  <WithFocusReturn(Component)>
                    <FocusReturn
                      childProps={
                        Object {
                          "children": <clickOutside(PopoverDetectOutside)>
                            <Animate
                              options={
                                Object {
                                  "origin": "top center",
                                }
                              }
                              type="appear"
                            >
                              [Function]
                            </Animate>
                          </clickOutside(PopoverDetectOutside)>,
                        }
                      }
                      focus={
                        Object {
                          "focusHistory": Array [],
                        }
                      }
                    >
                      <div
                        onBlur={[Function]}
                        onFocus={[Function]}
                      >
                        <Component>
                          <clickOutside(PopoverDetectOutside)>
                            <PopoverDetectOutside>
                              <Animate
                                options={
                                  Object {
                                    "origin": "top center",
                                  }
                                }
                                type="appear"
                              >
                                <ForwardRef
                                  className="components-popover is-bottom is-center is-mobile components-animate__appear is-from-top"
                                  onKeyDown={[Function]}
                                  style={
                                    Object {
                                      "left": undefined,
                                      "top": undefined,
                                      "visibility": undefined,
                                    }
                                  }
                                >
                                  <div
                                    className="components-popover is-bottom is-center is-mobile components-animate__appear is-from-top"
                                    onKeyDown={[Function]}
                                    onMouseDown={[Function]}
                                    style={
                                      Object {
                                        "left": undefined,
                                        "top": undefined,
                                        "visibility": undefined,
                                      }
                                    }
                                  >
                                    <div
                                      className="components-popover__header"
                                    >
                                      <span
                                        className="components-popover__header-title"
                                      />
                                      <ForwardRef(IconButton)
                                        className="components-popover__close"
                                        icon="no-alt"
                                      >
                                        <ForwardRef(Button)
                                          className="components-popover__close"
                                          icon="no-alt"
                                        >
                                          <button
                                            className="components-button components-popover__close has-icon"
                                            type="button"
                                          >
                                            <Icon
                                              icon="no-alt"
                                            >
                                              <Dashicon
                                                icon="no-alt"
                                                size={20}
                                              >
                                                <SVG
                                                  aria-hidden={true}
                                                  className="dashicon dashicons-no-alt"
                                                  focusable="false"
                                                  height={20}
                                                  role="img"
                                                  viewBox="0 0 20 20"
                                                  width={20}
                                                  xmlns="http://www.w3.org/2000/svg"
                                                >
                                                  <svg
                                                    aria-hidden="true"
                                                    className="dashicon dashicons-no-alt"
                                                    focusable="false"
                                                    height={20}
                                                    role="img"
                                                    viewBox="0 0 20 20"
                                                    width={20}
                                                    xmlns="http://www.w3.org/2000/svg"
                                                  >
                                                    <Path
                                                      d="M14.95 6.46L11.41 10l3.54 3.54-1.41 1.41L10 11.42l-3.53 3.53-1.42-1.42L8.58 10 5.05 6.47l1.42-1.42L10 8.58l3.54-3.53z"
                                                    >
                                                      <path
                                                        d="M14.95 6.46L11.41 10l3.54 3.54-1.41 1.41L10 11.42l-3.53 3.53-1.42-1.42L8.58 10 5.05 6.47l1.42-1.42L10 8.58l3.54-3.53z"
                                                      />
                                                    </Path>
                                                  </svg>
                                                </SVG>
                                              </Dashicon>
                                            </Icon>
                                          </button>
                                        </ForwardRef(Button)>
                                      </ForwardRef(IconButton)>
                                    </div>
                                    <div
                                      className="components-popover__content"
                                      style={
                                        Object {
                                          "maxHeight": undefined,
                                          "maxWidth": undefined,
                                        }
                                      }
                                      tabIndex="-1"
                                    >
                                      Hello
                                    </div>
                                  </div>
                                </ForwardRef>
                              </Animate>
                            </PopoverDetectOutside>
                          </clickOutside(PopoverDetectOutside)>
                        </Component>
                      </div>
                    </FocusReturn>
                  </WithFocusReturn(Component)>
                </div>
              </WithConstrainedTabbing(WithFocusReturn(Component))>
            </Portal>
          </FillComponent>
        </Fill>
        <ScrollLock />
      </span>
    </Popover>
  </PopoverWrapper>
  <Component>
    <Slot
      bubblesVirtually={true}
      name="Popover"
    >
      <SlotComponent
        bubblesVirtually={true}
        getFills={[Function]}
        name="Popover"
        registerSlot={[Function]}
        unregisterSlot={[Function]}
      >
        <div />
      </SlotComponent>
    </Slot>
  </Component>
</SlotFillProvider>
`;

exports[`renderPopover should render mobile popover 1`] = `
<SlotFillProvider>
  <WrapperComponent>
    <span>
      <Fill
        name="Popover"
      >
        <FillComponent
          name="Popover"
          registerFill={[Function]}
          unregisterFill={[Function]}
        >
          <Portal
            containerInfo={
              <div>
                <div
                  tabindex="-1"
                >
                  <div>
                    <div
                      class="components-popover is-middle is-center is-mobile is-without-arrow"
                      style="visibility: hidden;"
                    >
                      <div
                        class="components-popover__header"
                      >
                        <span
                          class="components-popover__header-title"
                        />
                        <button
                          class="components-button components-popover__close has-icon"
                          type="button"
                        >
                          <svg
                            aria-hidden="true"
                            class="dashicon dashicons-no-alt"
                            focusable="false"
                            height="20"
                            role="img"
                            viewBox="0 0 20 20"
                            width="20"
                            xmlns="http://www.w3.org/2000/svg"
                          >
                            <path
                              d="M14.95 6.46L11.41 10l3.54 3.54-1.41 1.41L10 11.42l-3.53 3.53-1.42-1.42L8.58 10 5.05 6.47l1.42-1.42L10 8.58l3.54-3.53z"
                            />
                          </svg>
                        </button>
                      </div>
                      <div
                        class="components-popover__content"
                        tabindex="-1"
                      />
                    </div>
                  </div>
                </div>
              </div>
            }
          >
            <WithConstrainedTabbing(WithFocusReturn(Component))>
              <div
                onKeyDown={[Function]}
                tabIndex="-1"
              >
                <WithFocusReturn(Component)>
                  <FocusReturn
                    childProps={
                      Object {
                        "children": <clickOutside(PopoverDetectOutside)
                          onClickOutside={[MockFunction]}
                        >
                          <Animate
                            options={
                              Object {
                                "origin": "middle center",
                              }
                            }
                            type={null}
                          >
                            [Function]
                          </Animate>
                        </clickOutside(PopoverDetectOutside)>,
                      }
                    }
                    focus={
                      Object {
                        "focusHistory": Array [],
                      }
                    }
                  >
                    <div
                      onBlur={[Function]}
                      onFocus={[Function]}
                    >
                      <Component>
                        <clickOutside(PopoverDetectOutside)
                          onClickOutside={[MockFunction]}
                        >
                          <PopoverDetectOutside
                            onClickOutside={[MockFunction]}
                          >
                            <Animate
                              options={
                                Object {
                                  "origin": "middle center",
                                }
                              }
                              type={null}
                            >
                              <ForwardRef
                                className="components-popover is-middle is-center is-mobile is-without-arrow"
                                onKeyDown={[Function]}
                                style={
                                  Object {
                                    "left": undefined,
                                    "top": undefined,
                                    "visibility": "hidden",
                                  }
                                }
                              >
                                <div
                                  className="components-popover is-middle is-center is-mobile is-without-arrow"
                                  onKeyDown={[Function]}
                                  onMouseDown={[Function]}
                                  style={
                                    Object {
                                      "left": undefined,
                                      "top": undefined,
                                      "visibility": "hidden",
                                    }
                                  }
                                >
                                  <div
                                    className="components-popover__header"
                                  >
                                    <span
                                      className="components-popover__header-title"
                                    />
                                    <ForwardRef(IconButton)
                                      className="components-popover__close"
                                      icon="no-alt"
                                      onClick={[MockFunction]}
                                    >
                                      <ForwardRef(Button)
                                        className="components-popover__close"
                                        icon="no-alt"
                                        onClick={[MockFunction]}
                                      >
                                        <button
                                          className="components-button components-popover__close has-icon"
                                          onClick={[MockFunction]}
                                          type="button"
                                        >
                                          <Icon
                                            icon="no-alt"
                                          >
                                            <Dashicon
                                              icon="no-alt"
                                              size={20}
                                            >
                                              <SVG
                                                aria-hidden={true}
                                                className="dashicon dashicons-no-alt"
                                                focusable="false"
                                                height={20}
                                                role="img"
                                                viewBox="0 0 20 20"
                                                width={20}
                                                xmlns="http://www.w3.org/2000/svg"
                                              >
                                                <svg
                                                  aria-hidden="true"
                                                  className="dashicon dashicons-no-alt"
                                                  focusable="false"
                                                  height={20}
                                                  role="img"
                                                  viewBox="0 0 20 20"
                                                  width={20}
                                                  xmlns="http://www.w3.org/2000/svg"
                                                >
                                                  <Path
                                                    d="M14.95 6.46L11.41 10l3.54 3.54-1.41 1.41L10 11.42l-3.53 3.53-1.42-1.42L8.58 10 5.05 6.47l1.42-1.42L10 8.58l3.54-3.53z"
                                                  >
                                                    <path
                                                      d="M14.95 6.46L11.41 10l3.54 3.54-1.41 1.41L10 11.42l-3.53 3.53-1.42-1.42L8.58 10 5.05 6.47l1.42-1.42L10 8.58l3.54-3.53z"
                                                    />
                                                  </Path>
                                                </svg>
                                              </SVG>
                                            </Dashicon>
                                          </Icon>
                                        </button>
                                      </ForwardRef(Button)>
                                    </ForwardRef(IconButton)>
                                  </div>
                                  <div
                                    className="components-popover__content"
                                    style={
                                      Object {
                                        "maxHeight": undefined,
                                        "maxWidth": undefined,
                                      }
                                    }
                                    tabIndex="-1"
                                  />
                                </div>
                              </ForwardRef>
                            </Animate>
                          </PopoverDetectOutside>
                        </clickOutside(PopoverDetectOutside)>
                      </Component>
                    </div>
                  </FocusReturn>
                </WithFocusReturn(Component)>
              </div>
            </WithConstrainedTabbing(WithFocusReturn(Component))>
          </Portal>
        </FillComponent>
      </Fill>
      <ScrollLock />
    </span>
  </WrapperComponent>
  <Component>
    <Slot
      bubblesVirtually={true}
      name="Popover"
    >
      <SlotComponent
        bubblesVirtually={true}
        getFills={[Function]}
        name="Popover"
        registerSlot={[Function]}
        unregisterSlot={[Function]}
      >
        <div />
      </SlotComponent>
    </Slot>
  </Component>
</SlotFillProvider>
`;

exports[`renderPopover should render popover 1`] = `
<SlotFillProvider>
  <WrapperComponent>
    <span>
      <Fill
        name="Popover"
      >
        <FillComponent
          name="Popover"
          registerFill={[Function]}
          unregisterFill={[Function]}
        >
          <Portal
            containerInfo={
              <div>
                <div
                  tabindex="-1"
                >
                  <div>
                    <div
                      class="components-popover is-middle is-center is-without-arrow"
                      style="top: 20px; left: 10px; visibility: hidden;"
                    >
                      <div
                        class="components-popover__content"
                        style="max-height: 30px; max-width: 40px;"
                        tabindex="-1"
                      />
                    </div>
                  </div>
                </div>
              </div>
            }
          >
            <WithConstrainedTabbing(WithFocusReturn(Component))>
              <div
                onKeyDown={[Function]}
                tabIndex="-1"
              >
                <WithFocusReturn(Component)>
                  <FocusReturn
                    childProps={
                      Object {
                        "children": <clickOutside(PopoverDetectOutside)
                          onClickOutside={[MockFunction]}
                        >
                          <Animate
                            options={
                              Object {
                                "origin": "middle center",
                              }
                            }
                            type={null}
                          >
                            [Function]
                          </Animate>
                        </clickOutside(PopoverDetectOutside)>,
                      }
                    }
                    focus={
                      Object {
                        "focusHistory": Array [],
                      }
                    }
                  >
                    <div
                      onBlur={[Function]}
                      onFocus={[Function]}
                    >
                      <Component>
                        <clickOutside(PopoverDetectOutside)
                          onClickOutside={[MockFunction]}
                        >
                          <PopoverDetectOutside
                            onClickOutside={[MockFunction]}
                          >
                            <Animate
                              options={
                                Object {
                                  "origin": "middle center",
                                }
                              }
                              type={null}
                            >
                              <ForwardRef
                                className="components-popover is-middle is-center is-without-arrow"
                                onKeyDown={[Function]}
                                style={
                                  Object {
                                    "left": "10px",
                                    "top": "20px",
                                    "visibility": "hidden",
                                  }
                                }
                              >
                                <div
                                  className="components-popover is-middle is-center is-without-arrow"
                                  onKeyDown={[Function]}
                                  onMouseDown={[Function]}
                                  style={
                                    Object {
                                      "left": "10px",
                                      "top": "20px",
                                      "visibility": "hidden",
                                    }
                                  }
                                >
                                  <div
                                    className="components-popover__content"
                                    style={
                                      Object {
                                        "maxHeight": "30px",
                                        "maxWidth": "40px",
                                      }
                                    }
                                    tabIndex="-1"
                                  />
                                </div>
                              </ForwardRef>
                            </Animate>
                          </PopoverDetectOutside>
                        </clickOutside(PopoverDetectOutside)>
                      </Component>
                    </div>
                  </FocusReturn>
                </WithFocusReturn(Component)>
              </div>
            </WithConstrainedTabbing(WithFocusReturn(Component))>
          </Portal>
        </FillComponent>
      </Fill>
    </span>
  </WrapperComponent>
  <Component>
    <Slot
      bubblesVirtually={true}
      name="Popover"
    >
      <SlotComponent
        bubblesVirtually={true}
        getFills={[Function]}
        name="Popover"
        registerSlot={[Function]}
        unregisterSlot={[Function]}
      >
        <div />
      </SlotComponent>
    </Slot>
  </Component>
</SlotFillProvider>
`;
